// @knapsack-pro/core/lib/ci-providers/buildkite.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Buildkite = void 0;
const _1 = require(".");
class Buildkite extends _1.CIProviderBase {
    static get ciNodeTotal() {
        return process.env.BUILDKITE_PARALLEL_JOB_COUNT;
    }
    static get ciNodeIndex() {
        return process.env.BUILDKITE_PARALLEL_JOB;
    }
    static get ciNodeBuildId() {
        return process.env.BUILDKITE_BUILD_NUMBER;
    }
    static get ciNodeRetryCount() {
        return process.env.BUILDKITE_RETRY_COUNT;
    }
    static get commitHash() {
        return process.env.BUILDKITE_COMMIT;
    }
    static get branch() {
        return process.env.BUILDKITE_BRANCH;
    }
    static get userSeat() {
        return (process.env.BUILDKITE_BUILD_AUTHOR || process.env.BUILDKITE_BUILD_CREATOR);
    }
    static get detect() {
        return 'BUILDKITE' in process.env ? this : null;
    }
    static get fixedQueueSplit() {
        return true;
    }
    static get ciProvider() {
        return 'Buildkite';
    }
}
exports.Buildkite = Buildkite;
