// @knapsack-pro/core/lib/ci-providers/ci-provider.base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CIProviderBase = void 0;
class CIProviderBase {
    static get ciNodeTotal() {
        throw new Error('nodeTotal getter is not implemented!');
    }
    static get ciNodeIndex() {
        throw new Error('nodeIndex getter is not implemented!');
    }
    static get ciNodeBuildId() {
        throw new Error('nodeBuildId getter is not implemented!');
    }
    static get ciNodeRetryCount() {
        throw new Error('ciNodeRetryCount getter is not implemented!');
    }
    static get commitHash() {
        throw new Error('commitHash getter is not implemented!');
    }
    static get branch() {
        throw new Error('branch getter is not implemented!');
    }
    static get userSeat() {
        throw new Error('userSeat getter is not implemented!');
    }
    static get fixedQueueSplit() {
        throw new Error('fixedQueueSplit getter is not implemented!');
    }
    static get ciProvider() {
        throw new Error('ciProvider getter is not implemented!');
    }
}
exports.CIProviderBase = CIProviderBase;
