// @knapsack-pro/core/lib/ci-providers/semaphore-ci.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SemaphoreCI = void 0;
const _1 = require(".");
class SemaphoreCI extends _1.CIProviderBase {
    static get ciNodeTotal() {
        return process.env.SEMAPHORE_THREAD_COUNT;
    }
    static get ciNodeIndex() {
        const currentThread = process.env.SEMAPHORE_CURRENT_THREAD;
        if (currentThread) {
            return (parseInt(currentThread, 10) - 1).toString();
        }
    }
    static get ciNodeBuildId() {
        return process.env.SEMAPHORE_BUILD_NUMBER;
    }
    static get ciNodeRetryCount() {
        return undefined;
    }
    static get commitHash() {
        return process.env.REVISION;
    }
    static get branch() {
        return process.env.BRANCH_NAME;
    }
    static get userSeat() {
        return undefined;
    }
    static get detect() {
        return 'SEMAPHORE_BUILD_NUMBER' in process.env ? this : null;
    }
    static get fixedQueueSplit() {
        return false;
    }
    static get ciProvider() {
        return 'Semaphore CI 1.0';
    }
}
exports.SemaphoreCI = SemaphoreCI;
