// @knapsack-pro/core/lib/ci-providers/semaphore-ci-2.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SemaphoreCI2 = void 0;
const _1 = require(".");
class SemaphoreCI2 extends _1.CIProviderBase {
    static get ciNodeTotal() {
        return process.env.SEMAPHORE_JOB_COUNT;
    }
    static get ciNodeIndex() {
        const jobIndex = process.env.SEMAPHORE_JOB_INDEX;
        if (jobIndex) {
            return (parseInt(jobIndex, 10) - 1).toString();
        }
    }
    static get ciNodeBuildId() {
        return process.env.SEMAPHORE_WORKFLOW_ID;
    }
    static get ciNodeRetryCount() {
        return undefined;
    }
    static get commitHash() {
        return process.env.SEMAPHORE_GIT_SHA;
    }
    static get branch() {
        return process.env.SEMAPHORE_GIT_BRANCH;
    }
    static get userSeat() {
        return undefined;
    }
    static get detect() {
        return 'SEMAPHORE' in process.env && 'SEMAPHORE_WORKFLOW_ID' in process.env
            ? this
            : null;
    }
    static get fixedQueueSplit() {
        return false;
    }
    static get ciProvider() {
        return 'Semaphore CI 2.0';
    }
}
exports.SemaphoreCI2 = SemaphoreCI2;
