// @knapsack-pro/core/lib/knapsack-pro-logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KnapsackProLogger = void 0;
const util = require("util");
const winston_1 = require("winston");
const config_1 = require("./config");
const { name: clientName } = require('../package.json');
class KnapsackProLogger {
    static objectInspect(object) {
        return util.inspect(object, {
            showHidden: false,
            depth: null,
            colors: true,
        });
    }
    logger;
    constructor(logLevel = config_1.KnapsackProEnvConfig.logLevel) {
        this.logger = (0, winston_1.createLogger)({
            level: logLevel,
            format: winston_1.format.combine(winston_1.format.label({ label: clientName }), winston_1.format.timestamp(), winston_1.format.colorize(), winston_1.format.printf(({ timestamp, label, level, message }) => `\n${timestamp} [${label}] ${level}: ${message}`)),
            transports: [new winston_1.transports.Console()],
        });
    }
    error(message) {
        this.logger.error(message);
    }
    warn(message) {
        this.logger.warn(message);
    }
    info(message) {
        this.logger.info(message);
    }
    verbose(message) {
        this.logger.verbose(message);
    }
    debug(message) {
        this.logger.debug(message);
    }
    silly(message) {
        this.logger.silly(message);
    }
}
exports.KnapsackProLogger = KnapsackProLogger;
